core/src/main/kotlin/model/Generics.kt

```diff
@@ -0,0 +1,135 @@
+package org.jetbrains.dokka.model
+
+import org.jetbrains.kotlin.descriptors.TypeParameterDescriptor
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.types.TypeProjection
+
+abstract class Type(open val name: String, open val bounds: List<Type>) {
+    open fun resolve(lookup: Map<String, GenericType.GenericTypeMutable>): Type = this
+
+    open fun asImmutable(
+        cache: MutableMap<String, GenericType>,
+        first: Boolean
+    ): Type = this
+
+    open fun genericCount(root: String): Int = 0
+    protected open fun setRef(cache: Map<String, GenericType>) {
+        bounds.filter { !(it is GenericType) || it is GenericType.GenericReference }.forEach { it.setRef(cache) }
+    }
+
+    abstract fun print(t: Type? = null): String
+}
+
+open class TypeConstructor(override val name: String, override val bounds: List<Type>) : Type(name, bounds) {
+    override fun genericCount(root: String): Int = 0
+    override fun toString(): String = "$name<${bounds.joinToString(separator = "") { "$it" }}>"
+    override fun print(t: Type?): String =
+        if (this == t) "" else "$name<${bounds.joinToString(separator = "") { it.print(t ?: this) }}>"
+}
+
+data class TypeConstructorMutable(override val name: String, override var bounds: List<Type>) :
```

Comment:
Data classes with vars are confusing. Does this even need to be declared as data class?